pkg/build/controller/controller.go

```diff
@@ -224,3 +224,71 @@ func (bc *BuildPodController) CancelBuild(build *buildapi.Build, pod *kapi.Pod)
 func isBuildCancellable(build *buildapi.Build) bool {
 	return build.Status == buildapi.BuildStatusNew || build.Status == buildapi.BuildStatusPending || build.Status == buildapi.BuildStatusRunning
 }
+
+// BuildPodDeleteController watches pods running builds and updates the build if the pod is deleted
+type BuildPodDeleteController struct {
+	BuildStore   cache.Store
+	BuildUpdater buildclient.BuildUpdater
+}
+
+func (bc *BuildPodDeleteController) HandleBuildPodDeletion(pod *kapi.Pod) error {
+	glog.V(4).Infof("Handling deletion of build pod %s/%s", pod.Namespace, pod.Name)
+	// pod and build have the same name, so look up the build by the pod's name.
+	obj, exists, err := bc.BuildStore.Get(pod)
+	if err != nil {
+		glog.V(4).Infof("Error getting build for pod %s/%s", pod.Namespace, pod.Name)
+		return err
+	}
+	if !exists || obj == nil {
+		glog.V(5).Infof("No build found for deleted pod %s/%s", pod.Namespace, pod.Name)
+		return nil
+	}
+	build := obj.(*buildapi.Build)
+
+	if build.Status != buildapi.BuildStatusRunning && build.Status != buildapi.BuildStatusPending && build.Status != buildapi.BuildStatusNew {
```

Comment:
I'd suggest we create some helper methods for getting information about different state of build, especially those that require you to check a couple of states. What do you think?